### PR TITLE
dai: use ibs/obs for dma buffer size calculation

### PR DESCRIPTION
**Pull request by iganakov**
Use ibs/obs size from ipc4_base_module_cfg to properly calculate period_count. It is especially important when FW aggregation mode is enabled and there are multiple DMAs allocated under one copier instance. That way period count for every DMA will be correctly evaluated and used for DMA buffer size calculation.
E.g. if `dma_buffer_size` received in Copier gtw config is 7680 bytes and FW should aggregate two streams 1ch/24_32bit/48kHz and 3ch/24_32bit/48kHz into one 4ch/24_32bit/48kHz stream, then FW will have to allocate two DMA's with sizes
1920 and 5760 bytes respectively and same 10 ms period for both to avoid glitches.

### CODE REVIEW

**Review by dabekjakub — APPROVED** [review]

File: src/audio/dai-zephyr.c [hunk 1]
```diff
@@ -848,12 +848,38 @@ static int dai_set_dma_buffer(struct dai_data *dd, struct comp_dev *dev,
...
+	 */
+	struct ipc4_copier_module_cfg *copier_cfg = dd->dai_spec_config;
+	uint32_t dma_buff_ms_size;
+
+	if (dev->direction == SOF_IPC_STREAM_CAPTURE)
+		dma_buff_ms_size = dd->ipc_config.dma_buffer_size / copier_cfg->base.ibs;
+	else
+		dma_buff_ms_size = dd->ipc_config.dma_buffer_size / copier_cfg->base.obs;
+
+	period_count = MAX(period_count, dma_buff_ms_size);
```
Comment (kv2019i) [inline src/audio/dai-zephyr.c:872]: Not quite following the logic. Period size may not always be 1ms, so how come we mix "period_count" and "dma_buff_ms_size" here? Or is the assumption that copier "ibs" must be aligned to period size of the pipeline. This would be good to capture in a comment. The variable name still needs to be corrected, it needs to be "dma_buff_size_in_periods" or something like that.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Some questions inline.

File: src/audio/dai-zephyr.c (hunk 1 above)
Comment (iganakov) [reply src/audio/dai-zephyr.c:872]: ibs/obs size depends on module processing domain, in case of Copier(LL) period will be 1ms. In Copier gtw config we receive DMA size aligned with ibs/obs. Based on previous assumptions we can calculate DMA buffer length in ms, which is also DMA `period_count`.  I agree the name of the variable and comment should be changed.

File: src/audio/dai-zephyr.c [hunk 2]
```diff
@@ -848,12 +848,38 @@ static int dai_set_dma_buffer(struct dai_data *dd, struct comp_dev *dev,
...
+	/* Use ibs/obs from copier gtw config to calculate DMA buffer length in ms.
+	 * This is used to calculate correct period count value.
+	 * It is especially important for FW aggregation use case, when Copier
+	 * receives whole DMA buffer size in GTW config but should be able to
+	 * allocate multiple DMA's implicitly.
+	 * E.g. if dma_buffer_size received in Copier GTW config is 7680 bytes and FW
+	 * should aggregate two streams 1ch/24_32bit/48kHz and 3ch/24_32bit/48kHz into one
+	 * 4ch/24_32bit/48kHz stream, then FW will have to allocate two DMA's with sizes
+	 * 1920 and 5760 bytes respectively and same 10 ms length (period_count) for both
+	 * to avoid glitches.
```
Comment (plbossart) [inline src/audio/dai-zephyr.c:862]: this explanation is very very hard to follow. For playback we typically send the same data to different links. For capture, we typically have a symmetry between the number of channels on different links.

The explanation with 1 and 3 channels does not describe which scenario you are trying to address, and whether this is an actual use case in a product or a test configuration.

In addition, the aggregation is at the link level, and one would hope that we don't have a 10ms buffer?  There's clearly a problem if we again have to rely on more buffers to get this solution to work properly...

**Review by plbossart — COMMENTED** [review]
not following this at all, sorry.

File: src/audio/dai-zephyr.c (hunk 2 above)
Comment (iganakov) [reply src/audio/dai-zephyr.c:862]: @plbossart there is no need for any additional buffers. I'm not going to change FW aggregation mechanism.
I found a bug in Windows production scenario (actual use case). When trying to aggregate two streams with different number of channels e.g. 1ch and 3ch (mono and 2.1) and 10ms DMA buffer requested from driver multiple glitches and first stream delay were observed. This change fixes that special case.

File: src/audio/dai-zephyr.c [hunk 3]
```diff
@@ -848,12 +848,38 @@ static int dai_set_dma_buffer(struct dai_data *dd, struct comp_dev *dev,
...
+	 */
+	struct ipc4_copier_module_cfg *copier_cfg = dd->dai_spec_config;
+	uint32_t dma_buff_length_ms;
+
+	if (dev->direction == SOF_IPC_STREAM_CAPTURE)
+		dma_buff_length_ms = dd->ipc_config.dma_buffer_size / copier_cfg->base.ibs;
+	else
+		dma_buff_length_ms = dd->ipc_config.dma_buffer_size / copier_cfg->base.obs;
+
+	period_count = MAX(period_count, dma_buff_length_ms);
```
Comment (plbossart) [inline src/audio/dai-zephyr.c:872]: Comparing period_count and milleseconds doesn't seem quite right. Maybe it works for the special case of a period being exactly one ms...

Comment (kv2019i) [reply src/audio/dai-zephyr.c:872]: @iganakov Can you modify the variable/comments here? This file is generic SOF code and the assumption of period to be 1ms only applies to current Intel IPC4 configurations. I.e. in the same function, we use "dev->frames" to calculate period size. It's very odd in the same function to assume a hardcoded period size later on. 

I'd propose:
```
uint32_t dma_buff_length_periods;

/* copier ibs/obs is set to size of one period */
if (dev->direction == SOF_IPC_STREAM_CAPTURE)
		dma_buff_length_periods = dd->ipc_config.dma_buffer_size / copier_cfg->base.ibs;
else
		dma_buff_length_periods = dd->ipc_config.dma_buffer_size / copier_cfg->base.obs;

period_count = MAX(period_count, dma_buff_length_periods);
```

**Review by kv2019i — COMMENTED** [review]
Left my proposal how to address the period/ms problem.

File: src/audio/dai-zephyr.c (hunk 3 above)
Comment (iganakov) [reply src/audio/dai-zephyr.c:872]: It seems reasonable, thanks @kv2019i
